src/Components/Components/src/NavigationManager.cs

```diff
@@ -274,6 +281,128 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
         }
     }
 
+    /// <summary>
+    /// Notifies the registered <see cref="IHandleLocationChanging"/> handlers of the current location change.
+    /// </summary>
+    /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI.</param>
+    /// <param name="isNavigationIntercepted">Whether this navigation was intercepted from a link.</param>
+    /// <param name="forceLoad">Whether the navigation is attempting to bypass client-side routing.</param>
+    /// <returns>A <see cref="ValueTask{TResult}"/> representing the completion of the operation.</returns>
+    public async ValueTask<bool> NotifyLocationChanging(string uri, bool isNavigationIntercepted, bool forceLoad)
+    {
+        _locationChangingCts?.Cancel();
+        _locationChangingCts = null;
+
+        if (_locationChangingHandlers.Count == 0)
+        {
+            return false;
+        }
+
+        _locationChangingCts = new();
+
+        var cancellationToken = _locationChangingCts.Token;
+        var context = new LocationChangingContext(uri, isNavigationIntercepted, forceLoad, cancellationToken);
+        var handlerCount = _locationChangingHandlers.Count;
+        var locationChangingHandlersCopy = ArrayPool<IHandleLocationChanging>.Shared.Rent(handlerCount);
+        _locationChangingHandlers.CopyTo(locationChangingHandlersCopy);
+
+        try
+        {
+            for (var i = 0; i < handlerCount; i++)
+            {
+                await locationChangingHandlersCopy[i].OnLocationChanging(context);
+
+                if (context.IsCanceled || cancellationToken.IsCancellationRequested)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+        catch (TaskCanceledException ex) when (ex.CancellationToken == cancellationToken)
+        {
+            // If it was our cancellation token, we should gracefully cancel the navigation.
+            return true;
+        }
+        catch (Exception ex)
+        {
+            throw new LocationChangeException("An exception occurred while dispatching a location changing event.", ex);
+        }
+        finally
+        {
+            ArrayPool<IHandleLocationChanging>.Shared.Return(locationChangingHandlersCopy);
+        }
+    }
+
+    /// <summary>
+    /// Sets whether there are active location changing handlers.
+    /// </summary>
+    /// <param name="value">Whether there are active location changing handlers.</param>
+    /// <returns><see langword="true"/> if the change was applied successfully, otherwise <see langword="false"/>.</returns>
```

Comment:
Is the intention that platform-specific subclasses should override this to perform a call into JS to enable/disable the lock?

If so a name like `SetNavigationLockState(bool)` might be clearer. The fact that the lock state is determined based on having a nonzero number of handlers can be a private matter for NavigationManager.

Apologies if I'm misunderstanding the flow here!

Reply:
Also consider having the base class throw `NotSupportedException` here, so as to make it really clear that subclasses have to override it (but only if this isn't going to break people's existing unit tests or such like - will this get called implicitly even for existing code that doesn't use the navigation lock functionality?).